### PR TITLE
[libc] Replace usage of GPU helpers with ones from 'gpuintrin.h'

### PR DESCRIPTION
**Pull request by jhuber6**
Summary:                                                                                                                                                                                                                              
These are provided by a resource header now, cut these from the                                                                                                                                                                       
dependencies and only provide the ones we use for RPC.

### CODE REVIEW

File: libc/src/__support/RPC/rpc.h [hunk 1]
```diff
@@ -30,6 +28,17 @@
 namespace LIBC_NAMESPACE_DECL {
 namespace rpc {
 
+/// Use scoped atomic variants if they are available for the target.
+#if !__has_builtin(__scoped_atomic_load_n)
+#define __scoped_atomic_load_n(src, ord, scp) __atomic_load_n(src, ord)
```
Comment (shiltian) [inline libc/src/__support/RPC/rpc.h:33]: you might want to guard all the arguments with `()`?

File: libc/src/__support/RPC/rpc.h [hunk 2]
```diff
@@ -190,28 +198,27 @@ template <bool Invert> struct Process {
...
-      atomic_thread_fence(cpp::MemoryOrder::ACQUIRE);
+      __atomic_thread_fence(__ATOMIC_ACQUIRE);
     return holding_lock;
   }
 
   /// Unlock the lock at index. We need a lane sync to keep this function
   /// convergent, otherwise the compiler will sink the store and deadlock.
-  [[clang::convergent]] LIBC_INLINE void unlock(uint64_t lane_mask,
-                                                uint32_t index) {
+  LIBC_INLINE void unlock(uint64_t lane_mask, uint32_t index) {
```
Comment (shiltian) [inline libc/src/__support/RPC/rpc.h:207]: These changes look irrelevant to this PR?

File: libc/src/__support/RPC/rpc.h (hunk 1 above)
Comment (jhuber6) [reply libc/src/__support/RPC/rpc.h:33]: These get undefed right after so I'm not too concerned about misuse.

File: libc/src/__support/RPC/rpc.h (hunk 2 above)
Comment (jhuber6) [reply libc/src/__support/RPC/rpc.h:207]: Just general cleanup, didn't feel worthy of its own PR.

File: libc/src/__support/RPC/rpc.h (hunk 1 above)
Comment (shiltian) [reply libc/src/__support/RPC/rpc.h:33]: Even for future-proofing, if someone else contributes to this and doesn’t realize it’s a macro or doesn’t use it correctly, I still believe it’s a good idea to implement it in a proper way. Not sure about the policy of libc project though. I'll leave it to others.

Comment (jhuber6) [reply libc/src/__support/RPC/rpc.h:33]: The arguments will be passed to the atomic function, what's a possible incorrect usage? I feel like even if it's malformed it'll be contained to the function argument scope.

File: libc/src/__support/RPC/rpc.h [hunk 3]
```diff
@@ -101,16 +110,16 @@ template <bool Invert> struct Process {
...
-        lane_mask,
-        inbox[index].load(cpp::MemoryOrder::RELAXED, cpp::MemoryScope::SYSTEM));
+    return rpc::broadcast_value(
+        lane_mask, __scoped_atomic_load_n(&inbox[index], __ATOMIC_RELAXED,
+                                          __MEMORY_SCOPE_SYSTEM));
   }
 
   /// Retrieve the outbox state from memory shared between processes.
   LIBC_INLINE uint32_t load_outbox(uint64_t lane_mask, uint32_t index) const {
-    return gpu::broadcast_value(lane_mask,
```
Comment (shiltian) [inline libc/src/__support/RPC/rpc.h:111]: IIUC, if we don't need `gpu::`, they will be removed right? I didn't find where they are removed.

Comment (jhuber6) [reply libc/src/__support/RPC/rpc.h:111]: Still used in the tests and other implementations like `printf` or `assert`. This is just trying to make the `rpc.h` header independent so we can put it somewhere else. Doing it bit by bit instead of all at once.

File: libc/src/__support/RPC/rpc.h [hunk 4]
```diff
@@ -67,18 +76,18 @@ template <bool Invert> struct Process {
   LIBC_INLINE ~Process() = default;
 
   uint32_t port_count = 0;
-  cpp::Atomic<uint32_t> *inbox = nullptr;
-  cpp::Atomic<uint32_t> *outbox = nullptr;
+  uint32_t *inbox = nullptr;
```
Comment (shiltian) [inline libc/src/__support/RPC/rpc.h:79]: Not sure which approach is better. With `std::atomic<>`, it’s clear that the variable is an atomic number. However, as a simple integer, its usage may not be safe if someone forgets to update it using atomic operations.

Comment (jhuber6) [reply libc/src/__support/RPC/rpc.h:79]: It's internal to the class, so we don't need to worry about anything external. If someone in the future modifies this I'll hope that they see that every access to this is atomic and that there's atomic in the comments.

File: libc/src/__support/RPC/rpc.h (hunk 2 above)
Comment (shiltian) [reply libc/src/__support/RPC/rpc.h:207]: This PR includes a fair number of cleanups. Idl open a separate PR specifically for those.

**Review by JonChesterfield — APPROVED** [review]
